compiler/dyno/include/chpl/framework/ErrorBase.h

```diff
@@ -0,0 +1,216 @@
+    if (kind_ != other.kind_ || type_ != other.type_) return false;
+    return contentsMatchInner(&other);
+  }
+  bool operator!=(const ErrorBase& other) const {
+    return !(*this == other);
+  }
+
+  virtual void write(ErrorWriter& wr) const = 0;
+  virtual void mark(Context* context) const = 0;
+};
+
+class ParseError : public ErrorBase {
+ private:
+  ID id_;
+  Location loc_;
+  std::string message_;
+
+ protected:
+  ParseError(Kind kind, ID id, Location loc, std::string message)
+    : ErrorBase(kind, PARSE), id_(std::move(id)),
+      loc_(std::move(loc)), message_(std::move(message)) {}
+
+  static const owned<ParseError>&
+  getParseError(Context* context, Kind kind, ID id, Location loc, std::string message);
+
+  bool contentsMatchInner(const ErrorBase* other) const override {
+    auto otherParse = static_cast<const ParseError*>(other);
+    return id_ == otherParse->id_ &&
+      loc_ == otherParse->loc_ &&
+      message_ == otherParse->message_;
+  }
+ public:
+
+  static const ParseError* get(Context* context, const ErrorMessage&);
+
+  void write(ErrorWriter& eq) const override;
+  void mark(Context* context) const override;
+};
+
+class GeneralError : public ErrorBase {
+ private:
+  ID id_;
+  Location loc_;
+  std::string message_;
+
+ protected:
+  GeneralError(ErrorBase::Kind kind, ID id, std::string message)
+    : ErrorBase(kind, GENERAL), id_(std::move(id)), message_(std::move(message)) {}
+
+  GeneralError(ErrorBase::Kind kind, Location loc, std::string message)
+    : ErrorBase(kind, GENERAL), loc_(std::move(loc)), message_(std::move(message)) {}
+
+  static const owned<GeneralError>&
+  getGeneralErrorID(Context* context, ErrorBase::Kind kind, ID id, std::string message);
+
+  static const owned<GeneralError>&
+  getGeneralErrorLocation(Context* context, ErrorBase::Kind kind, Location loc, std::string message);
+
+  bool contentsMatchInner(const ErrorBase* other) const override {
+    auto otherGeneral = static_cast<const GeneralError*>(other);
+    return id_ == otherGeneral->id_ &&
+      loc_ == otherGeneral->loc_ &&
+      message_ == otherGeneral->message_;
+  }
+ public:
+
+  static const GeneralError* vbuild(Context* context,
+                                    ErrorBase::Kind kind, ID id,
+                                    const char* fmt,
+                                    va_list vl);
+  static const GeneralError* vbuild(Context* context,
+                                    ErrorBase::Kind kind, Location loc,
+                                    const char* fmt,
+                                    va_list vl);
+
+  static const GeneralError* get(Context* context,
+                                 ErrorBase::Kind kind,
+                                 Location loc,
+                                 std::string msg);
+
+  void write(ErrorWriter& eq) const override;
+  void mark(Context* context) const override;
+};
+
+#define DIAGNOSTIC_CLASS(NAME, KIND, EINFO...)\
```

Comment:
Would appreciate a comment just before this along the lines of "Combine a #define and error-classes-list.h to produce class declarations for each error type. This way, the only steps needed to add an error are to update error-classes-list.h and to implement NewError::write."